twill-yarn/src/main/java/org/apache/twill/yarn/LocationCacheCleaner.java

```diff
@@ -0,0 +1,206 @@
+   */
+  private void cleanup(long currentTime) {
+    // First go through the pending cleanup list and remove those that can be removed
+    Iterator<PendingCleanup> iterator = pendingCleanups.iterator();
+    while (iterator.hasNext()) {
+      PendingCleanup pendingCleanup = iterator.next();
+
+      // If rejected by the predicate, it means it is being used, hence remove it from the pending cleanup list.
+      if (!cleanupPredicate.apply(pendingCleanup.getLocation())) {
+        iterator.remove();
+      } else {
+        try {
+          // If time is up for the pending entry, the location will be deleted,
+          // hence can be removed from the pending cleanup list.
+          // Otherwise retain it for the next cycle.
+          if (pendingCleanup.deleteIfExpired(currentTime)) {
+            iterator.remove();
+          }
+        } catch (IOException e) {
+          // Log and retain the entry so that another attempt on deletion will be made in next cleanup cycle
+          LOG.warn("Failed to delete {}", pendingCleanup.getLocation(), e);
+        }
+      }
+    }
+
+    // Then collects the next set of locations to be removed
+    try {
+      for (Location cacheDir : cacheBaseLocation.list()) {
+        try {
+          for (Location location : cacheDir.list()) {
+            if (cleanupPredicate.apply(location)) {
+              long expireTime = currentTime;
+              if (cacheDir.getName().equals(sessionId)) {
+                expireTime += expiry;
+              } else {
+                // If the cache entry is from different YarnTwillRunnerService session, use the anti expiry time.
+                expireTime += antiqueExpiry;
+              }
+              // If the location is already pending for cleanup, this won't update the expire time as
+              // the comparison of PendingCleanup is only by location.
+              pendingCleanups.add(new PendingCleanup(location, expireTime));
+            }
+          }
+        } catch (IOException e) {
+          LOG.warn("Failed to list cache content from {}", cacheDir, e);
+        }
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to list cache directories from {}", cacheBaseLocation, e);
+    }
+  }
+
+  /**
+   * Class for holding information about cache location that is pending to be removed.
+   * The equality and hash code is only based on the location.
+   */
+  private static final class PendingCleanup {
+    private final Location location;
+    private final long expireTime;
+
+    PendingCleanup(Location location, long expireTime) {
+      this.location = location;
+      this.expireTime = expireTime;
+    }
+
+    Location getLocation() {
+      return location;
+    }
+
+    long getExpireTime() {
+      return expireTime;
+    }
+
+    boolean deleteIfExpired(long currentTime) throws IOException {
+      if (currentTime < expireTime) {
+        return false;
+      }
+      location.delete();
```

Comment:
It is intentional to ignore the result of `delete()` call. When this method returns true, it means it is expired and delete was attempt so that it can be removed from the `pendingCleanups` set. Logically we don't care if the location actually be deleted successfully (if the location is already gone, we don't care. If fail to delete, it will get pick up again in the next cycle).

I'll add a log and also a comment to make this clear.